### PR TITLE
[TIMOB-26973] iOS: Add accessibilityIdentifier

### PR DESCRIPTION
**Pull request by drauggres**
https://jira.appcelerator.org/browse/TIMOB-26973

Compose  `accessibilityIdentifier` for views on iOS the same way `content description` is composed on Android.

### CODE REVIEW

**Comment by davidvivianoASICS** [comment]
Hello,

Is there an ETA of how long this PR will take to make it to production/next Titanium release?  accessibility-id's are something our team has been needing for QA automation and we are looking forward to their addition.

File: iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.h [hunk 1]
```diff
@@ -577,6 +577,8 @@ typedef enum {
 
 + (void)applyConstraintToView:(TiUIView *)view forProxy:(TiViewProxy *)proxy withBounds:(CGRect)bounds;
 
++ (NSString *)composeAccessibilityIdentifier:(TiUIView *)view;
```
Comment (vijaysingh-axway) [inline iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.h]: Can you change (TiUIView *) with (id) in above method declaration as TiUINavBarButton is not derived from TiUIView. So it is giving warning.

File: iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.m [hunk 2]
```diff
@@ -1420,6 +1420,51 @@ + (void)applyConstraintToView:(TiUIView *)view forProxy:(TiViewProxy *)proxy wit
   ApplyConstraintToViewWithBounds([proxy layoutProperties], view, bounds);
 }
 
++ (NSString *)composeAccessibilityIdentifier:(TiUIView *)view
```
Comment (vijaysingh-axway) [inline iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.m]: Same as above.

File: iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.m [hunk 3]
```diff
@@ -1420,6 +1420,51 @@ + (void)applyConstraintToView:(TiUIView *)view forProxy:(TiViewProxy *)proxy wit
...
+{
+  NSString *accessibilityLabel = view.accessibilityLabel;
```
Comment (vijaysingh-axway) [inline iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.m]: Will it always be unique? What is logic in this method to calculate accessibilityIdentifier.

**Review by vijaysingh-axway — CHANGES_REQUESTED** [review]
@drauggres  Can you please document the properties? And please address two minor comments as well. 
Can you give test case so that QE can verify this?
Thanks!

File: iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.h (hunk 1 above)
Comment (drauggres) [reply iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.h]: done

File: iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.m (hunk 2 above)
Comment (drauggres) [reply iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.m]: done

File: iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.m (hunk 3 above)
Comment (drauggres) [reply iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.m]: Not unique.
Logic is the same as for android `content description`: https://github.com/appcelerator/titanium_mobile/blob/af140652142b72b5e07eb5ed55964089ad2905c5/android/titanium/src/java/org/appcelerator/titanium/view/TiUIView.java#L2167

**Comment by drauggres** [comment]
```javascript
var win = Ti.UI.createWindow({
	backgroundColor: 'white',
	layout: 'vertical'
});

function getHex() {
	var a = (Math.random() * 256 | 0).toString(16);
	if (a.length < 2) {
		a = '0' + a;
	}
	return a;
}
function getRandomColor() {
	return '#' + getHex() + getHex() + getHex();
}

var view1 = Ti.UI.createView({
	backgroundColor: getRandomColor(),
	accessibilityLabel: 'view1Label',
	height: 10,
	width: Ti.UI.FILL
});

var view2 = Ti.UI.createView({
	backgroundColor: getRandomColor(),
	accessibilityValue: 'view3Value',
	height: 10,
	width: Ti.UI.FILL
});

var view3 = Ti.UI.createView({
	backgroundColor: getRandomColor(),
	accessibilityHint: 'view3Hint',
	height: 10,
	width: Ti.UI.FILL
});

var view4 = Ti.UI.createView({
	backgroundColor: getRandomColor(),
	accessibilityLabel: 'view4Label',
	accessibilityValue: 'view4Value',
	height: 10,
	width: Ti.UI.FILL
});

var view5 = Ti.UI.createView({
	backgroundColor: getRandomColor(),
	accessibilityLabel: 'view5Label',
	accessibilityHint: 'view5Hint',
	height: 10,
	width: Ti.UI.FILL
});

var view6 = Ti.UI.createView({
	backgroundColor: getRandomColor(),
	accessibilityValue: 'view6Value',
	accessibilityHint: 'view6Hint',
	height: 10,
	width: Ti.UI.FILL
});

var view7 = Ti.UI.createView({
	backgroundColor: getRandomColor(),
	accessibilityLabel: 'view7Label',
	accessibilityValue: 'view7Value',
	accessibilityHint: 'view7Hint',
	height: 10,
	width: Ti.UI.FILL
});

win.add(view1);
win.add(view2);
win.add(view3);
win.add(view4);
win.add(view5);
win.add(view6);
win.add(view7);

win.open();

```

appium-android
appium-ios

File: iphone/Classes/TiUINavBarButton.m [hunk 4]
```diff
@@ -124,6 +124,9 @@ - (id)initWithProxy:(TiUIButtonProxy *)proxy_
   proxy = proxy_; // Don't retain
 
   self.accessibilityLabel = [proxy_ valueForUndefinedKey:@"accessibilityLabel"];
+  self.accessibilityValue = [proxy_ valueForUndefinedKey:@"accessibilityValue"];
+  self.accessibilityHint = [proxy_ valueForUndefinedKey:@"accessibilityHint"];
+  self.accessibilityIdentifier = [TiUtils composeAccessibilityIdentifier:self];
```
Comment (vijaysingh-axway) [inline iphone/Classes/TiUINavBarButton.m]: TiUINavBarButton is neither derived from TiUIView nor UIAccessibilityElement. So accessibilityIdentifier will always be nil as per logic written in function 'composeAccessibilityIdentifier'. We can not make it nil.

**Review by vijaysingh-axway — CHANGES_REQUESTED** [review]
@drauggres  There is one minor issue. Please fix that.

**Review by vijaysingh-axway — APPROVED** [review]
CR passed.

**Comment by keerthi1032** [comment]
FR Passed. accessibility identifier is added. tested with test code provided in PR.works as expected

**Test Environment:**
Name                        = Mac OS X
  Version                     = 10.13.6
  Architecture                = 64bit
Node.js
  Node.js Version             = 8.9.1
  npm Version                 = 5.5.1
Titanium CLI
  CLI Version                 = 5.1.1
Titanium SDK
  SDK Version                 =local sdk  8.1.0.v20190515144948
Device =iPhone X iOS 11
Simulator =iphone XS iOS 12

**Comment by ygbr** [comment]
can we get this back-ported to 7.5.X?

**Pull request merged**